src/Temporalio/Bridge/CustomSlotSupplier.cs

```diff
@@ -0,0 +1,179 @@
+    /// <summary>
+    /// Core wrapper for a user-defined custom slot supplier.
+    /// </summary>
+    internal class CustomSlotSupplier : NativeInvokeableClass<Interop.CustomSlotSupplierCallbacks>
+    {
+        private readonly ILogger logger;
+        private readonly Temporalio.Worker.Tuning.ICustomSlotSupplier userSupplier;
+        private readonly Dictionary<uint, Temporalio.Worker.Tuning.ISlotPermit> permits = new();
+        private uint permitId = 1;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="CustomSlotSupplier" /> class.
+        /// </summary>
+        /// <param name="userSupplier">User's slot supplier implementation'.</param>
+        /// <param name="loggerFactory">Logger factory.</param>
+        internal unsafe CustomSlotSupplier(
+            Temporalio.Worker.Tuning.ICustomSlotSupplier userSupplier,
+            ILoggerFactory loggerFactory)
+        {
+            this.logger = loggerFactory.CreateLogger<CustomSlotSupplier>();
+            this.userSupplier = userSupplier;
+
+            var interopCallbacks = new Interop.CustomSlotSupplierCallbacks
+            {
+                reserve = FunctionPointer<Interop.CustomReserveSlotCallback>(Reserve),
+                cancel_reserve = FunctionPointer<Interop.CustomCancelReserveCallback>(CancelReserve),
+                try_reserve = FunctionPointer<Interop.CustomTryReserveSlotCallback>(TryReserve),
+                mark_used = FunctionPointer<Interop.CustomMarkSlotUsedCallback>(MarkUsed),
+                release = FunctionPointer<Interop.CustomReleaseSlotCallback>(Release),
+            };
+
+            PinCallbackHolder(interopCallbacks);
+        }
+
+        private static void SetCancelTokenOnCtx(ref SlotReserveCtx ctx, CancellationTokenSource cancelTokenSrc)
+        {
+            unsafe
+            {
+                try
+                {
+                    var handle = GCHandle.Alloc(cancelTokenSrc);
+                    fixed (Interop.SlotReserveCtx* p = &ctx)
+                    {
+                        Interop.Methods.set_reserve_cancel_target(p, GCHandle.ToIntPtr(handle).ToPointer());
+                    }
+                }
+                catch (Exception e)
+                {
+                    Console.WriteLine($"Error setting cancel token on ctx: {e}");
+                    throw;
+                }
+            }
+        }
+
+        private unsafe void Reserve(Interop.SlotReserveCtx ctx, void* sender)
+        {
+            SafeReserve(ctx, new IntPtr(sender));
+        }
+
+        private unsafe void CancelReserve(void* tokenSrc)
+        {
+            var handle = GCHandle.FromIntPtr(new IntPtr(tokenSrc));
+            var cancelTokenSrc = (CancellationTokenSource)handle.Target!;
+            cancelTokenSrc.Cancel();
+            handle.Free();
+        }
+
+        private void SafeReserve(Interop.SlotReserveCtx ctx, IntPtr sender)
+        {
+            var reserveTask = Task.Run(async () =>
```

Comment:
```suggestion
            Task.Run(async () =>
```

Were you getting an warning/error here if the variable wasn't assigned? If so, may be able to do `var _` or something (older versions of .NET won't support `_ =` maybe)